llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,309 @@
+        Replacements.push_back(Replacement(I, NewTrunc));  // Replace old instruction with new trunc for external users
+    }
+    else if (isa<BinaryOperator>(I)) {
+        BinaryOperator* BinOp = cast<BinaryOperator>(I);
+        bool NeedsPromotion = (BinOp->getType() == NonStdType);
+        
+        Value* LHS = getPromotedValue(BinOp->getOperand(0));
+        Value* RHS = getPromotedValue(BinOp->getOperand(1));
+        
+        Value* NewInst;
+        if (NeedsPromotion) {
+            // Create operation in promoted type
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        } else {
+            // For operations that should stay in original type
+            if (LHS->getType() != BinOp->getType())
+                LHS = Builder.CreateTrunc(LHS, BinOp->getType());
+            if (RHS->getType() != BinOp->getType())
+                RHS = Builder.CreateTrunc(RHS, BinOp->getType());
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        }
+        
+        errs() << Indent << "  " << *I << "    ============> " << *NewInst << "\n";
+        PromotedValues[I] = NewInst;
+        Replacements.push_back(Replacement(I, NewInst));
+    }
+    else if (isa<CallInst>(I)) {
+        CallInst* OldCall = cast<CallInst>(I);
+        // Create a new call with the same operands, but use promoted values where available
+        SmallVector<Value*, 8> NewArgs;
+        for (unsigned i = 0; i < OldCall->arg_size(); ++i) {
+            Value* OldArg = OldCall->getArgOperand(i);
+            Value* NewArg = PromotedValues.count(OldArg) ? PromotedValues[OldArg] : OldArg;
+            
+            // if the function expects a non-standard type, abort for now.
+            // TODO: if this assert is hit, we need to handle this case in the future
+            // by promoting the function arguments as well.
+            if (OldArg->getType() != NewArg->getType())
+                assert(false && "HipPromoteIntsPass: Function expects non-standard type");
+                
+            NewArgs.push_back(NewArg);
+        }
+        
+        CallInst* NewCall = CallInst::Create(OldCall->getFunctionType(),
+                                           OldCall->getCalledOperand(),
+                                           NewArgs,
+                                           OldCall->getName(),
+                                           OldCall);
+        NewCall->setCallingConv(OldCall->getCallingConv());
+        NewCall->setAttributes(OldCall->getAttributes());
+        
+        
+        errs() << Indent << "  " << *I << "    ============> " << *NewCall << "\n";
+        PromotedValues[I] = NewCall;
+        Replacements.push_back(Replacement(I, NewCall));
+    }
+    else {
+        assert(false && "HipPromoteIntsPass: Unhandled instruction type");
+    }
+}
+
+bool promoteChain(Instruction *OldI, Type *NonStdType, Type *PromotedTy,
```

Comment:
No caller observers the return value so might as change it to void. Also, this function does not need to be exported.
```suggestion
static void promoteChain(Instruction *OldI, Type *NonStdType, Type *PromotedTy,
```